src/meltano/core/project.py

```diff
@@ -77,26 +67,111 @@ def walk_parent_directories():
 
 
 class Project(Versioned):  # noqa: WPS214
-    """Represent the current Meltano project from a file-system perspective."""
+    """Represents a Meltano project."""
 
     __version__ = 1
     _activate_lock = threading.Lock()
     _find_lock = threading.Lock()
     _meltano_rw_lock = fasteners.ReaderWriterLock()
     _default = None
+    _cached_properties = (
+        "project_files",
+        "settings",
+        "plugins",
+    )
 
-    def __init__(self, root: os.PathLike):
-        """Instantiate a Project from its root directory.
+    def __init__(
+        self,
+        root: os.PathLike,
+        environment: Environment | None = None,
+        readonly: bool = False,
+    ):
+        """Initialize a `Project` instance.
 
         Args:
-            root: the root directory for the project
+            root: The root directory of the project.
+            environment: The active Meltano environment.
+            readonly: Whether the project is in read-only mode.
         """
         self.root = Path(root).resolve()
+        self.environment: Environment | None = environment
+        self.readonly = readonly
         self.sys_dir_root = Path(
-            os.getenv(PROJECT_SYS_DIR_ROOT, self.root / ".meltano")
+            os.getenv(PROJECT_SYS_DIR_ROOT_ENV, self.root / ".meltano")
         ).resolve()
-        self.readonly = False
-        self.active_environment: Environment | None = None
+
+    def refresh(self, **kwargs) -> None:
```

Comment:
Might be a stupid question but: what's the benefit of this approach vs. just making `Project` a singleton?

Reply:
Good question. Although it may not look like it, `Project` is effectively a singleton already because we retrieve it via `Project.find()` in almost all cases, and `Project.find()` caches the first project it returns (as `Project._default`), and then returns that instance until `Project.deactivate` is called.

Considering we practically have a singleton project object, and it keeps a `ProjectSettingsService`, `ProjectPluginsService`, etc., as attributes, the question becomes how can we make the project object reflect changes to the outside world (e.g. config updates). In the past we would just create a new `ProjectSettingsService`, `ProjectPluginsService`, etc., and they would obtain updates during their `__init__` methods. Now, since they are long-lived by default, we need to replace them with new instances. This `refresh` method accomplishes that.

Ideally I think we would have a read-only immutable `Project` object, and any update to the project would invalidate that instance, forcing you to create a new read-only view of the project. That would be a significantly larger refactor though, and I'm not ready to tackle something like that anytime soon.